addons/udes_advanced_packaging/models/stock_picking.py

```diff
@@ -0,0 +1,156 @@
+        in self. Where products_info is a dictionary mapped by
+        product ids and the value are the quantities.
+
+        The picking is also confirmed/assigned if the flags are set to True.
+        If result_package is set, it will update the result_package_id of the
+        new move_lines when assign flag is True.
+        """
+        Product = self.env["product.product"]
+        Move = self.env["stock.move"]
+        Package = self.env["stock.quant.package"]
+
+        self.assert_valid_state()
+
+        if values is None:
+            values = {}
+        values["picking_id"] = self.id
+        if not "location_id" in values:
+            values["location_id"] = self.location_id.id
+        if not "location_dest_id" in values:
+            values["location_dest_id"] = self.location_dest_id.id
+        if not "picking_type_id" in values:
+            values["picking_type_id"] = self.picking_type_id.id
+        if not "group_id" in values:
+            values["group_id"] = self.group_id.id
+        if not "product_uom" in values:
+            default_uom_id = self.env.ref("uom.product_uom_unit").id
+            values["product_uom"] = default_uom_id
+
+        for product_id, qty in products_info.items():
+            if product_quantity:
+                qty = product_quantity
+            if not qty:
+                raise ValidationError(
+                    _("No quantity specified creating move for product %s") % product_id
+                )
+            move_vals = {
+                "name": "{} {}".format(qty, Product.browse(product_id).display_name),
+                "product_id": product_id,
+                "product_uom_qty": qty,
+                "date": self.scheduled_date,
+            }
+            move_vals.update(values)
+            move = Move.create(move_vals)
+            if unexpected:
+                move.ordered_qty = 0
+
+        if confirm:
+            # Use picking.action_confirm, which will merge moves of the same
+            # product. In case that is not wanted use moves._action_confirm(merge=False)
+            # Use sudo() to allow unlink of the merged moves
+            self.sudo().action_confirm()
+
+        if assign:
+            old_move_line_ids = self.move_line_ids
+            # Use picking.action_assign or moves._action_assign to create move lines
+            # with context variable bypass_reservation_update in order to avoid
+            # to execute code specific for Odoo UI at stock.move.line.write()
+            self.with_context(bypass_reservation_update=True).action_assign()
+            if result_package:
+                # update result_package_id of the new move_line_ids
+                package = Package.get_package(result_package)
+                new_move_line_ids = self.move_line_ids - old_move_line_ids
+                new_move_line_ids.write({"result_package_id": package.id})
+
+    def assert_valid_state(self):
```

Comment:
NIt: valid state is ambiguos - is there a better name? assert_is_closed maybe?

Reply:
Fixed on latest commit